### PR TITLE
Issue 1536: (SegmentStore) Container de-registration sporadic issues.

### LINKED ISSUE #1536: Segment containers cannot be restarted immediately on start failures
**Problem description**
StreamSegmentContainerRegistry.startContainer() returns a completable future, which completes exceptionally on failure. But the containerid entry in segment registry is removed asynchronously. So immediate restart attempt on the same containerid returns a IllegalArgumentException since containerid is still seen to be registered.

**Problem location**
StreamSegmentContainerRegistry.startContainer()

**Suggestions for an improvement**
When startContainer() completable completes, the internal state should have already been updated so that subsequent starts could be attempted immediately if required.

### PR DESCRIPTION
**Pull request by andreipaduroiu**
**Change log description**
1. Fixed a bug in `StreamSegmentContainerRegistry` where it could erroneously fail a container registration request if that container has just been shut down. This was due to the fact that the internal de-registration happens in a callback which may be executed after the Future returned externally does. The solution provided handles all three possible cases: when a container fails during startup, when it shuts down (and results in failure or no failure) or when it fails while running.

2. Fixed a bug in `ServiceHelpers` where it could be possible that callbacks passed to onStop() would never execute in certain cases.

**Purpose of the change**
Fixes #1536.

**What the code does**
* `StreamSegmentContainerRegistry`: waits for a previous de-registration to happen before re-registering.
* `ServiceHelpers` handling a special case.

**How to verify it**
Unit tests added. Existing tests must pass.

### CODE REVIEW

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistry.java [hunk 1]
```diff
@@ -168,6 +186,14 @@ private void unregisterContainer(ContainerWithHandle containerWithHandle) {
...
         containerWithHandle.handle.notifyContainerStopped();
         log.info("Unregistered SegmentContainer {}.", containerWithHandle.handle.getContainerId());
+
+        containerWithHandle.shutdownNotifier.complete(null);
+    }
+
+    private static boolean isShutdown(Service.State state) {
+        return state == Service.State.FAILED
+                || state == Service.State.STOPPING
+                || state == Service.State.TERMINATED;
```
Comment (fpj) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistry.java]: This is not really a question about the changes here, but I was looking into understanding why we need a `TERMINATED` state. Is this a transient state until we get rid of the container instance? Also, to be consistent, shouldn't we call it `STOPPER` as the preceding state is `STOPPING`?

File: segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java [hunk 2]
```diff
@@ -207,28 +241,31 @@ public int getId() {
...
         }
 
         @Override
         protected void doStart() {
-            if (this.startException != null) {
-                notifyFailed(this.startException);
-            } else {
-                notifyStarted();
-            }
+            ForkJoinPool.commonPool().execute(() -> {
```
Comment (fpj) [inline segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java]: I'm not entirely sure why you are making this a ForkJoin task. Same for the stop call bellow.

File: segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java [hunk 3]
```diff
@@ -141,22 +144,52 @@ public void testContainerFailureWhileRunning() throws Exception {
...
+
+        registry.startContainer(containerId, TIMEOUT).join();
+        TestContainer container1 = (TestContainer) registry.getContainer(containerId);
+
+        // While already running.
+        AssertExtensions.assertThrows("startContainer() did not throw for already registered container.",
+                () -> registry.startContainer(containerId, TIMEOUT),
+                ex -> ex instanceof IllegalArgumentException);
+
+        // Immediately after a container fails - this should wait for any shutdown to occur and re-register it.
```
Comment (fpj) [inline segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java]: To confirm, this is optimistic, it is not guaranteed that `container1` won't have shut down by the time we call `startContainer` in line 186, right?

File: common/src/main/java/io/pravega/common/concurrent/ServiceHelpers.java [hunk 4]
```diff
@@ -71,32 +72,49 @@
      * @param executor           An Executor to use for callback invocations.
      */
     public static void onStop(Service service, Runnable terminatedCallback, Consumer<Throwable> failureCallback, Executor executor) {
+        ShutdownListener listener = new ShutdownListener(terminatedCallback, failureCallback);
+        service.addListener(listener, executor);
+
+        // addListener() will not invoke the callbacks if the service is already in a terminal state. As such, we need to
```
Comment (fpj) [inline common/src/main/java/io/pravega/common/concurrent/ServiceHelpers.java]: I don't follow this change compared to the description of the pull request. The description says:

```
Fixed a bug in ServiceHelpers where it could be possible that callbacks passed to onStop() would never execute in certain cases.
```

but the logic here compared to what is currently in master seems to guarantee no multiple calls to the callbacks, not that they will be invoked at least once. What am I missing?

File: segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java [hunk 5]
```diff
@@ -141,22 +144,52 @@ public void testContainerFailureWhileRunning() throws Exception {
...
+     */
+    @Test
+    public void testStartAlreadyRunning() throws Exception {
+        final int containerId = 1;
+        TestContainerFactory factory = new TestContainerFactory();
+        @Cleanup
+        StreamSegmentContainerRegistry registry = new StreamSegmentContainerRegistry(factory, executorService());
+
+        registry.startContainer(containerId, TIMEOUT).join();
+        TestContainer container1 = (TestContainer) registry.getContainer(containerId);
```
Comment (fpj) [inline segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java]: Minor: I'd rather avoid using numbers to suffix variable names. In this case, for example, you could have used something like: `containerToFail` and `containerToRestart`.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistry.java (hunk 1 above)
Comment (andreipaduroiu) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistry.java]: This is Guava's service state. TERMINATED == Stopped Successfully. FAILED == Stopped with failure.

File: segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java (hunk 2 above)
Comment (andreipaduroiu) [inline segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java]: Guava's Service doc instructs the `doStop`/`doStart` methods to execute their code async and return immediately, and only call `notifyStarted`/`notifyStopped`/`notifyFailed` when they are done. If not, then the whole start/stop would be synchronous. I just changed the unit test "Container" to execute stuff asynchronously - otherwise the condition that we are testing would never be triggered.

File: common/src/main/java/io/pravega/common/concurrent/ServiceHelpers.java (hunk 4 above)
Comment (andreipaduroiu) [inline common/src/main/java/io/pravega/common/concurrent/ServiceHelpers.java]: `addListener()` will attach a Listener - nothing more. If the service is already in a TERMINATED or FAILED state, the Listener will never get invoked. That was already handled before. However, it is possible that the service was not in one of these terminal states when we did the check, however, between that check and the time we register the Listener it would have transitioned to such a state - in that case the callback would have never gotten invoked.

This change moves the one-time check after the registration and also adds the code in the Listener to make sure that the callbacks are only invoked once - it is possible that the Service transitions to a terminal state after registering the Listener, but before the check, at which point the callbacks would be invoked twice.

File: segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java (hunk 3 above)
Comment (andreipaduroiu) [inline segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java]: Yes it's optimistic. It is possible that the async code actually executes before we are able to move on with the test, in which case it doesn't really test this corner case. However I noticed that this is hit pretty much every single time I run it (and I mean, I have run it in a loop 10K times and it hit every time).

I thought about making it deterministic, but that would require making this test container more complex. If you feel strongly about it I can add it.

File: segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java (hunk 5 above)
Comment (andreipaduroiu) [inline segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java]: Sure. I get your point. In this case there would be no issue in doing that. But I have other tests where I need to do more than two - in which case my inspiration runs out, and numbers make more sense (they also imply some sort of order).

File: segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java (hunk 2 above)
Comment (andreipaduroiu) [inline segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java]: Actually I just realized we are inheriting from `ThreadPooledTestSuite`. I changed this to use the provided Executor, and not the `ForkJoinPool`.

File: segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java (hunk 3 above)
Comment (andreipaduroiu) [inline segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistryTests.java]: I fixed it to be deterministic. I guess I was too lazy yesterday ...

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistry.java [hunk 6]
```diff
@@ -41,7 +41,7 @@
     //region Members
 
     private final SegmentContainerFactory factory;
-    private final AbstractMap<Integer, ContainerWithHandle> containers;
+    private final Map<Integer, ContainerWithHandle> containers;
```
Comment (tkaitchuck) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistry.java]: Perhaps ConcurrentMap?

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistry.java [hunk 7]
```diff
@@ -150,6 +143,31 @@ public SegmentContainer getContainer(int containerId) throws ContainerNotFoundEx
 
     //region Helpers
 
+    private CompletableFuture<ContainerHandle> startContainerInternal(int containerId) {
+        // Create a new Container and attempt to register it. Be optimistic about it: create it first and then attempt
```
Comment (tkaitchuck) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistry.java]: Please turn these into a method doc

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistry.java (hunk 6 above)
Comment (andreipaduroiu) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistry.java]: Why is it important? The implementation itself is a ConcurrentHashMap?

Comment (andreipaduroiu) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistry.java]: fixed.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistry.java (hunk 7 above)
Comment (andreipaduroiu) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/StreamSegmentContainerRegistry.java]: fixed
